File: api/src/models/QuestionModel.py
from typing import List, TYPE_CHECKING
import uuid
from src.models.AnswerModel import Answer
from src.database import Base
from sqlalchemy import Boolean, CheckConstraint, Column, ForeignKey, Integer, String, Date, UUID
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import Mapped, mapped_column, relationship


if TYPE_CHECKING:
    from src.models.SurveyModel import Survey
    from src.models.OptionModel import Option


class Question(Base):
    __tablename__ = "questions"



    id: Mapped[uuid.UUID] = mapped_column(UUID, primary_key=True, index=True, default=uuid.uuid4)
    number: Mapped[int] = mapped_column(Integer, nullable=False)
    description: Mapped[str] = mapped_column(String(250), nullable=False)
    type: Mapped[str] = mapped_column(String(100), nullable=False)
    required: Mapped[bool] = mapped_column(Boolean, nullable=False)
    survey_id: Mapped[uuid.UUID] = mapped_column(UUID, ForeignKey("surveys.id"), nullable=False)



    survey: Mapped["Survey"] = relationship(back_populates="questions", lazy="selectin")
    options: Mapped[List["Option"]] = relationship(back_populates="question", cascade="all, delete", lazy="selectin")
    answers: Mapped[List["Answer"]] = relationship(back_populates="question", cascade="all, delete", lazy="selectin")


    __table_args__ = (
        CheckConstraint("type IN ('single_choice', 'multiple_choice', 'likert_scale', 'open')", name="type_check"),
        CheckConstraint("number > 0", name="number_check"),
    )

File: client/surbased/lib/src/survey/domain/question_model.dart
import 'package:surbased/src/survey/domain/option_model.dart';

class Question {
  final int? number;
  final String? description;
  final String? type;
  final bool? required;
  final String? id;
  final String? surveyId;
  final List<Option>? options;
  String? text;

  Question({
    this.number,
    this.description,
    this.type,
    this.required,
    this.id,
    this.surveyId,
    this.options,
    this.text,
  });

  factory Question.fromJson(Map<String, dynamic> json) => Question(
        number: json["number"],
        description: json["description"],
        type: json["type"],
        required: json["required"],
        id: json["id"],
        surveyId: json["survey_id"],
        options: json["options"] == null 
            ? [] 
            : List<Option>.from(json["options"].map((x) => Option.fromJson(x))),
        text: json["text"],
      );

  Map<String, dynamic> toJson() {
     final Map<String, dynamic> json = {};
     if (id != null) json["id"] = id;
     if (surveyId != null) json["survey_id"] = surveyId;
     if (number != null) json["number"] = number;
     if (description != null) json["description"] = description;
     if (type != null) json["type"] = type;
     if (required != null) json["required"] = required;
     if (options != null) {
       json["options"] = List<dynamic>.from(options!.map((x) => x.toJson()));
     }
     if (text != null) json["text"] = text;
     return json;
   }
}
